README.md
# aeson-picker

[![Travis](https://img.shields.io/travis/ozzzzz/aeson-picker.svg)](https://travis-ci.org/ozzzzz/aeson-picker)
[![hackage](https://img.shields.io/hackage/v/aeson-picker.svg)](https://hackage.haskell.org/package/aeson-picker)
[![hackage-deps](https://img.shields.io/hackage-deps/v/aeson-picker.svg)](https://hackage.haskell.org/package/aeson-picker)

Tiny library to get fields from JSON format

Common use is the following:
```
JSON |-- FIELDS :: EXPECTED TYPE
Text |-- [Text] :: a
```
So operator `(|--)` gets JSON (represented as `Text`), "route" to field inside JSON (represented as `[Text]`), and tries to parse field from JSON with described route to expected type.
If expected type can be deduced with type checker then it can be dropped down.

A little bit safer operator is `(|-?)`. It returns not `a` but `Maybe a`.

## Examples
First, add extension (to not pack `String` to `Text` every time):
```
ghci>:set -XOverloadedStrings
```

Then you can try something simple (empty list means that you try to parse `Value` from the top JSON-level):
```
ghci>"5" |-- [] :: Int
5
ghci>"5" |-- [] :: Float
5.0
ghci>"5" |-- [] :: String
"*** Exception: Data.Aeson.Picker: could not pick field with path: []
```

But what if field you are looking for somewhere inside JSON?
That's why are you here.

Let's try to get something from inside JSON:
```
ghci>"{\"a\": 5}" |-- ["a"] :: Int
5
```
But be sure that the field is presented inside JSON:
```
ghci>"{\"a\": 5}" |-- ["b"] :: Int
*** Exception: Data.Aeson.Picker: could not pick field with path: ["b"]
```
We can go deeper (as deep as you want):
```
ghci>"{\"outer\": {\"inner\": [1,2,3]}}" |-- ["outer", "inner"] :: [Int]
[1,2,3]
```
But be sure that you JSON is really valid (by specification key in JSON should be `String`):
```
ghci>"{a: 5}" |-- ["a"] :: Int
*** Exception: Data.Aeson.Picker: input json is not valid
```

If you want more "safe" picker, you can use another operator:
```
ghci>"5" |-? [] :: Maybe Int
Just 5
ghci>"{\"a\": 5}" |-? ["a"] :: Maybe Int
Just 5
ghci>"{\"a\": 5}" |-? ["b"] :: Maybe Int
Nothing
```

In current logic even operator `(|-?)` will throw error if JSON is not valid:
```
ghci>"{a: 5}" |-? ["a"] :: Maybe Int
*** Exception: Data.Aeson.Picker: input json is not valid
```

You can open issue if you do not think that it is right logic.



CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [0.1.0.6] - 2022-03-23
### Changed
- Support aeson-2 and other GHC-9 related packages.

## [0.1.0.5] - 2019-09-07
### Changed
- Upper bounds for lens-aeson.

## [0.1.0.0] - 2018-01-22
### Added
- Initial version for picking.


test/Spec.hs
{-# LANGUAGE OverloadedStrings #-}

import           Data.Aeson.Picker
import           Data.Text         (Text, unpack)
import           Test.Hspec

exampleJSON :: Text
exampleJSON = "{\"object-like\":{\"string-like\":\"this is string\", \"int-like\": 42}, \"float-like\":1.2}"

main :: IO ()
main = hspec $ do
    describe ("With JSON " ++ unpack exampleJSON) $ do
      it "|-- [\"float-like\"]" $
           (exampleJSON |-- ["float-like"]) `shouldBe` (1.2 :: Float)
      it "|-- [\"object-like\", \"string-like\"]" $
           (exampleJSON |-- ["object-like", "string-like"]) `shouldBe` ("this is string" :: String)
      it "|-- [\"object-like\", \"int-like\"]" $
           (exampleJSON |-- ["object-like", "int-like"]) `shouldBe` (42 :: Int)
      
     
      
     


src/Data/Aeson/Picker/Internal/Functions.hs
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE RankNTypes        #-}

module Data.Aeson.Picker.Internal.Functions
  ( (|--)
  , (|-?)
  ) where

import           Control.Lens    (Traversal', (^?))
import           Data.Aeson      (FromJSON (..), Result (..), Value, fromJSON)
import           Data.Aeson.Lens (AsValue, key, _Value)
import           Data.Maybe      (fromMaybe)
import           Data.Text       (Text)

{- |
From given JSON and selectors returns typed field. If input JSON is not valid or selected field is not found then error is thrown.
If you need more safe way use '(|-?)' instead. Examples:

-- >>> "5" |-- [] :: Int
-- 5
-- >>> "5" |-- [] :: Float
-- 5.0
-- >>> "5" |-- [] :: String
-- *** Exception: Data.Aeson.Picker: could not pick field with path: []
--
-- >>> :set -XOverloadedStrings
-- >>> "{\"a\": 5}" |-- ["a"] :: Int
-- 5
-- >>> "{\"a\": 5}" |-- ["b"] :: Int
-- *** Exception: Data.Aeson.Picker: could not pick field with path: ["b"]
-- >>> "{\"outer\": {\"inner\": [1,2,3]}}" |-- ["outer", "inner"] :: [Int]
-- [1,2,3]
-- >>> {\"outer\": {\"inner\": [1,2,3]}}" |-- ["outer", "inner"] :: [Double]
-- [1.0,2.0,3.0]
-- >>> "{a: 5}" |-- ["a"] :: Int
-- *** Exception: Data.Aeson.Picker: input json is not valid
-}
infix 5 |--
(|--) :: (AsValue t, FromJSON a) => t -> [Text] -> a
json |-- selectors = fromMaybe (error $ "Data.Aeson.Picker: could not pick field with path: " ++ show selectors) $ json |-? selectors

{- |
From given JSON and selectors returns typed field inside 'Maybe'. If input JSON is not valid then error is thrown.
Examples:

-- >>> "5" |-? [] :: Maybe Int
-- Just 5
-- >>> "5" |-? [] :: Maybe String
-- Nothing
-- >>> "{\"a\": 5}" |-? ["a"] :: Maybe Int
-- Just 5
-- >>> "{a: 5}" |-? ["a"] :: Maybe Int
-- *** Exception: Data.Aeson.Picker: input json is not valid
-}
infix 5 |-?
(|-?) :: (AsValue t, FromJSON a) => t -> [Text] -> Maybe a
json |-? selectors = let validJSON = checkValidity json
                     in pick validJSON selectors >>= convert

-- | Checks validity for JSON format. Throws error if it is not valid.
checkValidity :: AsValue t => t -> t
checkValidity json = fromMaybe (error "Data.Aeson.Picker: input json is not valid") (json ^? _Value) `seq` json

-- | Picks from given JSON selected field
pick :: AsValue t => t -> [Text] -> Maybe Value
pick json []        = json ^? _Value
pick json selectors = json ^? genGetter selectors

-- | Converts from 'Value'
convert :: FromJSON a => Value -> Maybe a
convert value = case fromJSON value of
  Success r -> Just r
  Error _   -> Nothing

-- | Generates getter from given selectors
genGetter :: AsValue t => [Text] -> Traversal' t Value
genGetter []     = error "Data.Aeson.Picker.Internal.Functions.genGetter: this should not be happened"
genGetter [x]    = key x
genGetter (x:xs) = key x . genGetter xs




src/Data/Aeson/Picker.hs
{-# LANGUAGE CPP               #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE RankNTypes        #-}
module Data.Aeson.Picker
  ( (|--)
  , (|-?)
  ) where

import           Control.Lens    (Traversal', (^?))
import           Data.Aeson      (FromJSON (..), Result (..), Value, fromJSON)
import           Data.Aeson.Lens (AsValue, key, _Value)
import           Data.Maybe      (fromMaybe)
import           Data.Text       (Text)

#if MIN_VERSION_aeson(2, 0, 0)
import qualified Data.Aeson.Key
#endif

-- | From given JSON and selectors returns typed field. If input JSON is not valid or selected field is not found then error is thrown.
-- If you need more safe way use '(|-?)' instead. Examples:
--
-- > ghci> "5" |-- [] :: Int
-- > 5
-- > ghci> "5" |-- [] :: Float
-- > 5.0
-- > ghci>"5" |-- [] :: String
-- > *** Exception: Data.Aeson.Picker: could not pick field with path: []
-- >
-- > ghci> :set -XOverloadedStrings
-- > ghci> "{\"a\": 5}" |-- ["a"] :: Int
-- > 5
-- > ghci> "{\"a\": 5}" |-- ["b"] :: Int
-- > *** Exception: Data.Aeson.Picker: could not pick field with path: ["b"]
-- > ghci> "{\"outer\": {\"inner\": [1,2,3]}}" |-- ["outer", "inner"] :: [Int]
-- > [1,2,3]
-- > ghci> {\"outer\": {\"inner\": [1,2,3]}}" |-- ["outer", "inner"] :: [Double]
-- > [1.0,2.0,3.0]
-- > ghci> "{a: 5}" |-- ["a"] :: Int
-- > *** Exception: Data.Aeson.Picker: input json is not valid
infix 5 |--
(|--) :: (AsValue t, FromJSON a) => t -> [Text] -> a
json |-- selectors = fromMaybe (error $ "Data.Aeson.Picker: could not pick field with path: " ++ show selectors) $ json |-? selectors

-- | From given JSON and selectors returns typed field inside 'Maybe'. If input JSON is not valid then error is thrown.
-- Examples:
--
-- > ghci> "5" |-? [] :: Maybe Int
-- > Just 5
-- > ghci> "5" |-? [] :: Maybe String
-- > Nothing
-- > ghci> "{\"a\": 5}" |-? ["a"] :: Maybe Int
-- > Just 5
-- > ghci> "{a: 5}" |-? ["a"] :: Maybe Int
-- > *** Exception: Data.Aeson.Picker: input json is not valid
infix 5 |-?
(|-?) :: (AsValue t, FromJSON a) => t -> [Text] -> Maybe a
json |-? selectors = let validJSON = checkValidity json
                     in pick validJSON selectors >>= convert

-- | Checks validity for JSON format. Throws error if it is not valid.
checkValidity :: AsValue t => t -> t
checkValidity json = fromMaybe (error "Data.Aeson.Picker: input json is not valid") (json ^? _Value) `seq` json

-- | Picks from given JSON selected field
pick :: AsValue t => t -> [Text] -> Maybe Value
pick json []        = json ^? _Value
pick json selectors = json ^? genGetter selectors

-- | Converts from 'Value'
convert :: FromJSON a => Value -> Maybe a
convert value = case fromJSON value of
  Success r -> Just r
  Error _   -> Nothing

-- | Generates getter from given selectors
genGetter :: AsValue t => [Text] -> Traversal' t Value
genGetter []     = error "Data.Aeson.Picker.Internal.Functions.genGetter: this should not be happened"
#if MIN_VERSION_aeson(2, 0, 0)
genGetter [x]    = key (Data.Aeson.Key.fromText x)
genGetter (x:xs) = key (Data.Aeson.Key.fromText x) . genGetter xs
#else
genGetter [x]    = key x
genGetter (x:xs) = key x . genGetter xs
#endif
